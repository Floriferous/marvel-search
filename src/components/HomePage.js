import React from 'react';

import Input from './Input';
import CharacterList from './CharacterList';
import Disclaimer from './Disclaimer';

const HomePage = () => (
  <main>
    <h1>Marvel Search</h1>
    <Input />
    <CharacterList />
    <Disclaimer />
  </main>
);

export default HomePage;
