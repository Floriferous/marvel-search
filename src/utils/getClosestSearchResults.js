import { createSearchKey } from '../actions/search';

/**
 * getClosestSearchResults - Searches for the closest search results
 * currently in memory. It iterates on a search string, and removes one
 * character at a time until it finds an old search result to display.
 *
 * If the first run passes, it simply returns the exact search results.
 *
 * An isLoading flag is also returned, to indicate whether old or exact
 * results are currently being returned
 *
 * @param {Object} searchResults Object with all previous search results
 * @param {String} search        current search string
 * @param {String} pagination    current pagination
 *
 * @return {Object}
 */
const getClosestSearchResults = (searchResults, search, pagination) => {
  let broaderResults;
  let isExactSearch;

  // Return immediately if search length equals 1 or pagination is larger than 0
  if (search.length === 1 || pagination > 0) {
    return {
      data: searchResults[createSearchKey(search, pagination)] || {},
      isLoading: false, // taken care of by CharacterList
    };
  }

  for (let i = 0; i < search.length; i += 1) {
    const broaderSearch = search.slice(0, search.length - i);
    broaderResults = searchResults[createSearchKey(broaderSearch, pagination)];
    if (broaderResults) {
      isExactSearch = i === 0;
      break;
    }
  }
  return { data: broaderResults || {}, isLoading: !isExactSearch };
};

export default getClosestSearchResults;
