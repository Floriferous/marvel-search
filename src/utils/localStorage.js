export const storageAvailable = () => {
  try {
    const storage = localStorage;
    const x = '__storage_test__';
    storage.setItem(x, x);
    storage.removeItem(x);
    return true;
  } catch (e) {
    return false;
  }
};

export const saveValueAtKey = (key, value) => {
  if (storageAvailable()) {
    localStorage.setItem(key, JSON.stringify(value));
    return true;
  }

  return false;
};

export const getValueAtKey = (key) => {
  if (storageAvailable()) {
    return JSON.parse(localStorage.getItem(key));
  }

  return false;
};
