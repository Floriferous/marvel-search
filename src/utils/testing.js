import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

export const testComponent = (Component, props) => {
  if (!testComponent.mountedLockScreen) {
    testComponent.mountedLockScreen = Enzyme.shallow(<Component {...props} />);
  }
  return testComponent.mountedLockScreen;
};

testComponent.reset = () => {
  testComponent.mountedLockScreen = undefined;
};

export const { shallow, mount, render } = Enzyme;
export default Enzyme;
