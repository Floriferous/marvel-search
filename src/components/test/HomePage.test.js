import React from 'react';
import renderer from 'react-test-renderer';

import { shallow } from '../../utils/testing';
import HomePage from '../HomePage';

describe('HomePage', () => {
  let component;

  beforeEach(() => {
    component = () => shallow(<HomePage />);
  });

  // it('matches its snapshot', () => {
  //   const tree = renderer.create(<HomePage />).toJSON();
  //   expect(tree).toMatchSnapshot();
  // });

  it('renders a main as its root', () => {
    expect(component().find('main').length).toBe(1);
  });

  it('renders one h1, Input and CharacterList', () => {
    expect(component().find('h1').length).toBe(1);
    expect(component().find('Input').length).toBe(1);
    expect(component().find('CharacterList').length).toBe(1);
  });
});
