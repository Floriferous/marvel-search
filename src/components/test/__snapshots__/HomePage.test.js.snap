// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage matches its snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <HomePage />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Marvel Search
</h1>,
        <Connect(Component) />,
        <Connect(Component) />,
        <Disclaimer />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Marvel Search",
        },
        "ref": null,
        "rendered": "Marvel Search",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "main",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Marvel Search
</h1>,
          <Connect(Component) />,
          <Connect(Component) />,
          <Disclaimer />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Marvel Search",
          },
          "ref": null,
          "rendered": "Marvel Search",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "main",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
    },
  },
}
`;
